### ISSUE TITLE
Mojo::Reactor::EV: Timer failed

### ISSUE THREAD

**Issue by karel-m** [kraih/mojo#1247]
* Mojolicious version: 7.88
* Perl version: 5.20.1
* Operating system: Linux / CentOS 6 / x86_64

```
CORE
  Perl        (v5.20.1, linux)
  Mojolicious (7.88, Doughnut)

OPTIONAL
  EV 4.0+                 (4.22)
  IO::Socket::Socks 0.64+ (n/a)
  IO::Socket::SSL 2.009+  (2.056)
  Net::DNS::Native 0.15+  (0.18)
  Role::Tiny 2.000001+    (2.000006)
```

### Actual behavior

Basically very similar behaviour to what I have described in issue #1235. At certain point in time the application hits some I/O corner case and falls into an unstable state emitting errors like:

```
## this is approx. 4 minutes long fraction of relevant log messages:
[error] Can't call method "isa" on an undefined value at [..]/Mojo/IOLoop/Stream/HTTPClient.pm line 25.
[warn] Mojo::Reactor::EV: Timer failed: Connection already closed at /usr/local/perl520/site/lib/Mojolicious/Plugin/DefaultHelpers.pm line 134.
[warn] Mojo::Reactor::EV: Timer failed: Connection already closed at /usr/local/perl520/site/lib/Mojolicious/Plugin/DefaultHelpers.pm line 134.
[error] Can't call method "isa" on an undefined value at [..]/Mojo/IOLoop/Stream/HTTPClient.pm line 25.
[error] Can't call method "isa" on an undefined value at [..]/Mojo/IOLoop/Stream/HTTPClient.pm line 25.
[warn] Mojo::Reactor::EV: Timer failed: Connection already closed at /usr/local/perl520/site/lib/Mojolicious/Plugin/DefaultHelpers.pm line 134.
[error] Can't call method "isa" on an undefined value at [..]/Mojo/IOLoop/Stream/HTTPClient.pm line 25.
[warn] Mojo::Reactor::EV: Timer failed: Connection already closed at /usr/local/perl520/site/lib/Mojolicious/Plugin/DefaultHelpers.pm line 134.
[warn] Mojo::Reactor::EV: Timer failed: Connection already closed at /usr/local/perl520/site/lib/Mojolicious/Plugin/DefaultHelpers.pm line 134.
[error] Can't call method "isa" on an undefined value at [..]/Mojo/IOLoop/Stream/HTTPClient.pm line 25.
[error] Can't call method "isa" on an undefined value at [..]/Mojo/IOLoop/Stream/HTTPClient.pm line 25.
[warn] Mojo::Reactor::EV: Timer failed: Connection already closed at /usr/local/perl520/site/lib/Mojolicious/Plugin/DefaultHelpers.pm line 134.
[error] Can't call method "isa" on an undefined value at [..]/Mojo/IOLoop/Stream/HTTPClient.pm line 25.
[warn] Mojo::Reactor::EV: Timer failed: Connection already closed at /usr/local/perl520/site/lib/Mojolicious/Plugin/DefaultHelpers.pm line 134.
```

Currently I do not have a full stack trace yet.

I have put `use Carp::Always;` to `Reactor/EV.pm` and `Reactor/Poll.pm` so hopefully I'll get more information during the next error occurrence.

### Expected behavior

On pre-7.83 the same applications code run without any error like this (or similar).

### Steps to reproduce the behavior

In #1235 I wrote that "This is the hardest part". Here the things are much worse as this failure seems to occur on our production site approx once in 7-10 days.

**Comment by kraih** [comment]
Yea, this seems impossible to debug. I don't think there is anything we can do.

**Comment by kraih** [comment]
I'll assume the fix for #1235 was wrong and have reverted it. https://github.com/kraih/mojo/commit/bdc185c8611dd4047b4297a41a17389fde02d9c5

**Comment by kraih** [comment]
I will not be trying any more hacks to fix the problem, since it turned out so badly last time. So i'm afraid you will have to track down the problem yourself, or provide a test case we can use to replicate the problem.

**Comment by karel-m** [comment]
This time the fail-cycle was a bit faster, here is the stack trace:

```
[Tue Aug  7 16:29:01 2018] [warn] "27660 Mojo::Reactor::EV: Timer failed: Connection already closed at [..]/lib/Mojolicious/Controller.pm line 230.
#### the rest is our code #### 
Mojolicious::Plugin::LogHash::__ANON__(Mojo::UserAgent=HASH(0xd347438), Mojo::Transaction::HTTP=HASH(0xde19970)) called at [..]/lib/Mojo/UserAgent.pm line 237
Mojo::UserAgent::_finish(Mojo::UserAgent=HASH(0xd347438), \"ac30aa2db27e8490cbe123e275a8cf80\") called at [..]/lib/Mojo/UserAgent.pm line 249
Mojo::UserAgent::__ANON__(Mojo::Transaction::HTTP=HASH(0xde19970)) called at [..]/lib/Mojo/EventEmitter.pm line 15
Mojo::EventEmitter::emit(Mojo::Transaction::HTTP=HASH(0xde19970), \"finish\") called at [..]/lib/Mojo/Transaction.pm line 16
Mojo::Transaction::closed(Mojo::Transaction::HTTP=HASH(0xde19970)) called at [..]/lib/Mojo/UserAgent.pm line 225
Mojo::UserAgent::_error(Mojo::UserAgent=HASH(0xd347438), \"ac30aa2db27e8490cbe123e275a8cf80\", \"Inactivity timeout\") called at [..]/lib/Mojo/UserAgent.pm line 137
Mojo::UserAgent::__ANON__(Mojo::IOLoop::Stream::HTTPClient=HASH(0xde2a9b0)) called at [..]/lib/Mojo/EventEmitter.pm line 15
Mojo::EventEmitter::emit(Mojo::IOLoop::Stream::HTTPClient=HASH(0xde2a9b0), \"timeout\") called at [..]/lib/Mojo/IOLoop/Stream.pm line 77
Mojo::IOLoop::Stream::__ANON__(Mojo::Reactor::EV=HASH(0x1fb4458)) called at [..]/lib/Mojo/Reactor/Poll.pm line 145
eval {...} called at [..]/lib/Mojo/Reactor/Poll.pm line 145
Mojo::Reactor::Poll::_try(Mojo::Reactor::EV=HASH(0x1fb4458), \"Timer\", CODE(0xdfea150)) called at [..]/lib/Mojo/Reactor/EV.pm line 66
Mojo::Reactor::EV::__ANON__(EV::Timer=SCALAR(0xddd2b18), 256) called at [..]/lib/Mojo/Reactor/EV.pm line 27
eval {...} called at [..]/lib/Mojo/Reactor/EV.pm line 27
Mojo::Reactor::EV::start(Mojo::Reactor::EV=HASH(0x1fb4458)) called at [..]/lib/Mojo/IOLoop.pm line 140
Mojo::IOLoop::start(Mojo::IOLoop=HASH(0x435f948)) called at [..]/lib/Mojo/Server/Prefork.pm line 163
Mojo::Server::Prefork::_spawn(Mojo::Server::Prefork=HASH(0x1c519a8)) called at [..]/lib/Mojo/Server/Prefork.pm line 100
Mojo::Server::Prefork::_manage(Mojo::Server::Prefork=HASH(0x1c519a8)) called at [..]/lib/Mojo/Server/Prefork.pm line 85
Mojo::Server::Prefork::run(Mojo::Server::Prefork=HASH(0x1c519a8)) called at [..]/lib/Mojo/Server/Hypnotoad.pm line 74
Mojo::Server::Hypnotoad::run(Mojo::Server::Hypnotoad=HASH(0x1c51708), \"/opt/website/app/Portal/script/asq_portal\") called at [..]/bin/hypnotoad line 14"
        Mojo::Reactor::Poll::_try(Mojo::Reactor::EV=HASH(0x1fb4458), "Timer", CODE(0xe2ac110)) called at [..]/lib/Mojo/Reactor/EV.pm line 66
        Mojo::Reactor::EV::__ANON__(EV::Timer=SCALAR(0xe298390), 256) called at [..]/lib/Mojo/Reactor/EV.pm line 27
```

I'll try to investigate more.

**Comment by kraih** [comment]
Mojolicious 7.89 has been released with the reverted change. So you can use that to have it fail much faster.

**Comment by kraih** [comment]
@fskale How is that relevant to this problem?

**Comment by fskale** [comment]
When using TLS and IO::Socket::SSL < 2.0.58 is in place, the memleak will fill up the memory until the OOM will start to kill processes which can be a subprocess of your app and of course, when in the middle of a transaction, the master process notices as the forked process left the building prior on closing the connection. Since i have busy servers, i debugged it and found the problem which is not a big deal when using a monitoring system. Dmesg showed the OOM events and then i took a look at the bug reports at the IO::Socket::SSL github site. As i use the CRL feature i found another bug with regards to not free the crl structure.

**Comment by kraih** [comment]
@fskale I don't see anything hinting at subprocesses being involved here. What am i missing?

**Comment by Corion** [comment]
I have a very similar problem with another backend, the ::Poll backend, on Windows and Linux. I get the crash as:
```
Mojo::Reactor::Poll: I/O watcher failed: Can't call method "isa" on an undefined
 value at C:/strawberry-perl-5.26.0.1-64bit/perl/site/lib/Mojo/IOLoop/Stream/HTTPClient.pm line 25.
```

This seems to be a regression between 7.43 (my old Mojolicious version) and 
Mojolicious 7.89. CPAN testers have the same problem with the test suite of Future::HTTP:

https://www.cpantesters.org/distro/F/Future-HTTP.html?oncpan=1&distmat=1&version=0.09&grade=3

The test t/01-mojo.t fails in 100% of the cases with Mojolicious 7.89.

**Comment by fskale** [comment]
Without a dmesg, i cannot tell, but you hit the bug, when using IO::Socket:SSL and a busy server, since every connection will leave an unfreed mem of approx 124kb. OOM doesn't necessearilly kill the master but a preforked child which can have an open connection when OOM kills it. It would be great to have the kernel version and dmesg, memory usage (htop) etc. details.
I had a similar problem (TLS) and it vanished after upgrading to 2.0.58 though.
Anyhow, is use 
<pre>

   CORE
  Perl        (v5.20.2, linux)
  Mojolicious (7.89, Doughnut)
  Linux: Debian Jessie (kernel version 4.1.11)

OPTIONAL
  Cpanel::JSON::XS 4.04+  (4.04)
  EV 4.0+                 (4.22)
  IO::Socket::Socks 0.64+ (0.65)
  IO::Socket::SSL 2.009+  (2.058)
  Net::DNS::Native 0.15+  (0.15)
  Role::Tiny 2.000001+    (2.000006)
</pre>
and cannot replicate the problem on a really busy server.

**Comment by kraih** [comment]
@Corion Thanks, i've been able to track it down with your test to a connection close race condition. This patch seems to fix it.
```perl
diff --git a/lib/Mojo/UserAgent.pm b/lib/Mojo/UserAgent.pm
index 80f641a24..917a190a9 100644
--- a/lib/Mojo/UserAgent.pm
+++ b/lib/Mojo/UserAgent.pm
@@ -271,7 +271,7 @@ sub _reuse {
   my $c   = $self->{connections}{$id};
   my $tx  = delete $c->{tx};
   my $max = $self->max_connections;
-  return $self->_remove($id) if $close || !$tx || !$max;
+  return $self->_remove($id) if $close || !$tx || !$max || !$tx->keep_alive;

   # Keep connection alive
   my $queue = $self->{queue}{$c->{ioloop}} ||= [];
```

**Comment by kraih** [comment]
The fix has been committed. https://github.com/kraih/mojo/commit/2865f3b101e630d192d6eca8352171d028ad7be6

**Comment by karel-m** [comment]
I have manually patched `Mojo/UserAgent.pm` on our trouble making site according to sri's patch above.

More details: 
* TLS is not involved at all (so it is unlikely to be related to IO::Socket::SSL)
* we have updated all perl modules to their latest CPAN versions (as of today)
* the kernel is exactly what the latest CentOS 6 (today: 6.10) has as the latest kernel (today: 2.6.32-754.2.1.el6.x86_64)

As mentioned in my previous issue #1235 our simplified architecture looks like:
```
[browser]---->[nginx]---->[FE-mojo]---->[BE-mojo]---->[PgSQL]
                              ^
                              |
                           TROUBLE!
```

Basically every request from a browser causes that `FE-mojo` makes at least one call via `app->ua` (Mojo::UserAgent) to `BE-mojo`. And after some time `app->ua` objects on `FE-mojo` end up in the unstable/error state described at the beginning of this issue.

Thank you for your time.

**Comment by kraih** [comment]
@karel-m If i had to guess i'd say the race condition happens sometimes when BE-mojo reaches the end of a keep-alive connection and sends `Connection: close` back to FE-mojo.

**Comment by kraih** [comment]
Mojolicious 7.90 has been released with the fix.

**Comment by fskale** [comment]
Quite old kernel,
are you sure, your're not hitting the EPOLL_WAIT Bug (kernel before 2.6.37 ?)
man epoll_wait:
<pre>
BUGS
In  kernels before 2.6.37, a timeout value larger than approximately
 LONG_MAX / HZ milliseconds is reated as -1 (i.e., infinity).  Thus,
 for example, on a system where the sizeof(long) is 4 and the kernel 
HZ value is 1000, this means that timeouts greater than 35.79 
minutes  are  treated  as infinity.
</pre>

**Comment by karel-m** [comment]
As for the kernel we believe that RedHat guys know what they are doing.  We simply use the latest CentOS 6 (= RHEL6) kernel. AFAIK it is not vanilla 2.6.32; it is 2.6.32 + some stuff that RedHat considered worth backporting.

**Comment by kraih** [comment]
I don't see how anything @fskale said has any relevance here. We know what cause the problem and fixed it already.
